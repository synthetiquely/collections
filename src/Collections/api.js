export default [
  {
    id: '1',
    src:
      'http://chert-poberi.ru/wp-content/uploads/proga/111/images1/201712/igor5-17121714072507_5.jpg',
    title: 'A kitty with a cup of coffee',
  },
  {
    id: '2',
    src: 'https://uc.lolkot.ru/picture/47363-37992-3.jpg',
    title: 'A kitty with a cup of coffee',
  },
  {
    id: '3',
    src:
      'http://s10.favim.com/orig/161123/animal-cat-coffee-kitty-Favim.com-4889090.jpeg',
    title: 'A kitty with a cup of coffee',
  },
  {
    id: '4',
    src:
      'http://scontent-arn2-1.cdninstagram.com/t51.2885-15/s640x640/sh0.08/e35/22430171_134521763938677_674873313641627648_n.jpg',
    title: 'A kitty with a cup of coffee',
  },
  {
    id: '5',
    src:
      'https://78.media.tumblr.com/87d2659d6345c1e230c77081b8af003b/tumblr_ouc1k3npX61sojkzpo1_1280.jpg',
    title: 'A kitty with a cup of coffee',
  },
  {
    id: '6',
    src:
      'http://cdn2.arhivurokov.ru/viewImage.php?image=http://img0.liveinternet.ru/images/attach/d/0/138/501/138501004_132018492_tumblr_n3nvvjICtT1tv66qwo1_1280.jpg',
    title: 'A kitty with a cup of coffee',
  },
];
