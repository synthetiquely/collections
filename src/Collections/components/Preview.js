import React from 'react';
import glamorous from 'glamorous';

import Modal from '../../Common/Modal';

const Image = glamorous.img({
  width: '100%',
  height: '100%',
  objectFit: 'cover',
});

export default ({ src, title, ...rest }) => (
  <Modal {...rest}>
    <Image src={src} alt={title} />
  </Modal>
);
