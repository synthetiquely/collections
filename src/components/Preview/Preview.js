import React, { Component } from 'react';
import { observer } from 'mobx-react';
import Modal from '../Modal/Modal';
import Swipe from '../Swipe/Swipe';
import Image from '../styled/Image';
import Tooltip from '../styled/Tooltip';
import TooltipText from '../styled/TooltipText';

import {
  DESTINATION_PREVIOUS,
  DESTINATION_NEXT,
  KEY_ARROW_LEFT,
  KEY_ARROW_RIGHT,
  KEY_ESC,
} from '../../constants';

@observer
class Preview extends Component {
  constructor(props) {
    super(props);
    this.state = { showTooltip: false, loaded: false };
    this.onLoad = this.onLoad.bind(this);
    this.onSwipe = this.onSwipe.bind(this);
    this.onMouseMove = this.onMouseMove.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.onClose = this.onClose.bind(this);
    this.onChangeSelected = this.onChangeSelected.bind(this);
  }

  componentDidMount() {
    document.body.addEventListener('keydown', this.onKeyDown);
  }

  componentDidUpdate() {
    if (this.props.collections.selectedPhoto) {
      document.body.classList.add('overlayed');
    } else {
      document.body.classList.remove('overlayed');
    }
  }

  componentWillUnmount() {
    document.body.removeEventListener('keydown', this.onKeyDown);
  }

  onLoad() {
    this.setState({
      loaded: true,
    });
  }

  onChangeSelected(destination) {
    this.props.collections.slideNext(destination);
  }

  onClose() {
    this.props.collections.clearSelection();
    this.setState({
      loaded: false,
      showTooltip: false,
    });
  }

  onKeyDown(e) {
    if (e.keyCode) {
      // eslint-disable-next-line
      switch (e.keyCode) {
        case KEY_ARROW_LEFT:
          this.onChangeSelected(DESTINATION_PREVIOUS);
          break;
        case KEY_ARROW_RIGHT:
          this.onChangeSelected(DESTINATION_NEXT);
          break;
        case KEY_ESC:
          this.onClose();
      }
    }
  }

  onMouseMove() {
    this.setState(prevState => ({
      showTooltip: !prevState.showTooltip,
    }));
  }

  onSwipe(direction) {
    if (direction === 'left' || direction === 'down') {
      this.onChangeSelected(DESTINATION_PREVIOUS);
    } else if (direction === 'right' || direction === 'up') {
      this.onChangeSelected(DESTINATION_NEXT);
    }
  }

  render() {
    const { showTooltip, loaded } = this.state;

    if (!this.props.collections.selectedPhoto) {
      return null;
    }

    return (
      <Modal
        bgColor={!loaded && this.props.collections.selectedPhoto.color}
        {...this.props}
        onClose={this.onClose}
        onChangeSelected={this.onChangeSelected}
      >
        <Swipe onSwipe={this.onSwipe}>
          <a href={this.props.collections.selectedPhoto.url} target="_blank">
            <Image
              style={{ width: 'auto', maxWidth: '100%', objectFit: 'contain' }}
              loaded={loaded}
              src={this.props.collections.selectedPhoto.fullsrc}
              alt={this.props.collections.selectedPhoto.description}
              onLoad={this.onLoad}
              onMouseEnter={this.onMouseMove}
              onMouseLeave={this.onMouseMove}
            />
            {loaded && (
              <Tooltip id="preview-tooltip" showTooltip={showTooltip && loaded}>
                <TooltipText>
                  {this.props.collections.selectedPhoto.user.username}
                </TooltipText>
              </Tooltip>
            )}
          </a>
        </Swipe>
      </Modal>
    );
  }
}

export default Preview;
