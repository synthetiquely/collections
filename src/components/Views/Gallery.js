import glamorous from 'glamorous';

const Gallery = glamorous.section(
  {
    display: 'flex',
    flexWrap: 'wrap',
    '@media only screen and (min-width: 768px)': {
      '&::after': {
        content: "''",
        flexGrow: '999999999',
      },
    },
  },
  ({ isOpen }) => ({
    overflowY: isOpen ? 'hidden' : 'auto',
  }),
);

export default Gallery;
