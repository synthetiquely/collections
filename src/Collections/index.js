import React, { Component } from 'react';
import glamorous from 'glamorous';

import ImagePreview from './components/ImagePreview';
import Image from './components/Image';
import api from './api';
// import calculateImageOrientation from '../utils/imageUtils';

class Collections extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: null,
    };
    this.onClose = this.onClose.bind(this);
    this.onNext = this.onNext.bind(this);
    this.onPrevious = this.onPrevious.bind(this);
  }

  onClick(index) {
    this.setState({
      selected: index,
    });
  }

  onNext(e) {
    e.stopPropagation();
    e.nativeEvent.stopImmediatePropagation();
    if (api[this.state.selected + 1]) {
      this.setState(prevState => ({
        selected: prevState.selected + 1,
      }));
    } else {
      this.setState({
        selected: 0,
      });
    }
  }

  onPrevious(e) {
    e.stopPropagation();
    e.nativeEvent.stopImmediatePropagation();
    if (api[this.state.selected - 1]) {
      this.setState(prevState => ({
        selected: prevState.selected - 1,
      }));
    } else {
      this.setState({
        selected: api.length - 1,
      });
    }
  }

  onClose(e) {
    console.log('Closed', e.target);

    this.setState({
      selected: null,
    });
  }

  renderItems() {
    // @FIXME: Refactor transofrmation pipeline with rxJS when redux will be available
    return (
      api
        // .map(item => ({
        //   ...item,
        //   orientation: calculateImageOrientation(item.src),
        // }))
        .map((item, index) => (
          <Image
            src={item.src}
            title={item.title}
            orientation={item.orientation || 'even'}
            key={item.id}
            onClick={() => this.onClick(index)}
          />
        ))
    );
  }

  render() {
    const { selected } = this.state;
    return (
      <Views isOpen={selected !== null}>
        {this.renderItems()}
        {selected !== null && (
          <ImagePreview
            isOpen={selected !== null}
            src={api[selected].src}
            id={api[selected].id}
            title={api[selected].title}
            onClose={this.onClose}
            onNext={this.onNext}
            onPrevious={this.onPrevious}
          />
        )}
      </Views>
    );
  }
}

const Views = glamorous.section(
  {
    display: 'grid',
    gridGap: '10px',
    gridTemplateColumns: 'repeat(auto-fit, minmax(100px, 1fr))',
    gridAutoRows: '75px',
    gridAutoFlow: 'dense',
    padding: '10px',
  },
  ({ isOpen }) => ({
    overflowY: isOpen ? 'hidden' : 'auto',
  }),
);

export default Collections;
