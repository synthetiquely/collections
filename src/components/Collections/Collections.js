import React, { Component } from 'react';
import glamorous from 'glamorous';

import InfiniteScroll from '../InfiniteScroll/InfiniteScroll';
import Preview from '../Preview/Preview';
import View from '../View/View';
import * as constants from '../../constants';
import { imagesLoaded } from '../../utils/imageUtils';

class Collections extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: null,
      loading: true,
    };
    this.galleryElement = null;
    this.onClose = this.onClose.bind(this);
    this.onChangeLoading = this.onChangeLoading.bind(this);
    this.onChangeSelected = this.onChangeSelected.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.setRef = this.setRef.bind(this);
  }

  componentDidMount() {
    document.addEventListener('keydown', this.onKeyDown);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown);
  }

  onClick(index) {
    this.setState({
      selected: index,
    });
  }

  onKeyDown(e) {
    if (this.state.selected !== null && e.keyCode) {
      switch (e.keyCode) {
        case constants.KEY_ARROW_LEFT:
          this.onChangeSelected(null, constants.DESTINATION_PREVIOUS);
          break;
        case constants.KEY_ARROW_RIGHT:
          this.onChangeSelected(null, constants.DESTINATION_NEXT);
          break;
      }
    }
  }

  onChangeSelected(e, destination) {
    if (e) {
      e.stopPropagation();
      e.nativeEvent.stopImmediatePropagation();
    }

    const { images } = this.props;
    if (destination === constants.DESTINATION_NEXT) {
      if (images[this.state.selected + 1]) {
        this.setState(prevState => ({
          selected: prevState.selected + 1,
        }));
      } else {
        this.setState({
          selected: 0,
        });
      }
    } else if (destination === constants.DESTINATION_PREVIOUS) {
      if (images[this.state.selected - 1]) {
        this.setState(prevState => ({
          selected: prevState.selected - 1,
        }));
      } else {
        this.setState({
          selected: images.length - 1,
        });
      }
    }
  }

  onChangeLoading() {
    const isLoaded = !imagesLoaded(this.galleryElement);

    this.setState({
      loading: isLoaded,
    });
  }

  onClose() {
    this.setState({
      selected: null,
    });
  }

  setRef(element) {
    this.galleryElement = element;
  }

  renderItems() {
    if (!this.props.images) {
      return null;
    }
    return this.props.images.map((item, index) => (
      <View
        src={item.src}
        title={item.title}
        key={item.id}
        loaded={!this.state.loading}
        onClick={() => this.onClick(index)}
        onChangeLoading={this.onChangeLoading}
      />
    ));
  }

  render() {
    const { selected } = this.state;
    const { images } = this.props;

    return (
      <InfiniteScroll>
        <Views innerRef={this.setRef} isOpen={selected !== null}>
          {this.renderItems()}
          {selected !== null && (
            <Preview
              src={images[selected].src}
              id={images[selected].id}
              title={images[selected].title}
              onClose={this.onClose}
              onChangeSelected={this.onChangeSelected}
            />
          )}
        </Views>
      </InfiniteScroll>
    );
  }
}

const Views = glamorous.section(
  {
    display: 'flex',
    flexWrap: 'wrap',
    '@media only screen and (min-width: 768px)': {
      '&::after': {
        content: "''",
        flexGrow: '999999999',
      },
    },
  },
  ({ isOpen }) => ({
    overflowY: isOpen ? 'hidden' : 'auto',
  }),
);

export default Collections;
