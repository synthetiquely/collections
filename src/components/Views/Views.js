import React from 'react';
import View from '../View/View';
import Gallery from './Gallery';

const Views = ({ images, isOpen, onClick }) => {
  if (!images) {
    return null;
  }

  const renderItems = () =>
    images.map((item, index) => (
      <View
        src={item.src}
        title={item.title}
        key={item.id}
        onClick={onClick(index)}
      />
    ));

  return <Gallery isOpen={isOpen}>{renderItems()}</Gallery>;
};

export default Views;
