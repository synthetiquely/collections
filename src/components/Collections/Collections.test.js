import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { Provider } from 'mobx-react';
import Collections from './Collections';
import Views from '../Views/Views';
import Preview from '../Preview/Preview';

const collections = {
  photos: [
    {
      id: '1',
      description: '',
      color: '#444',
      width: '150',
      height: '216',
      links: {
        html: 'example.com',
      },
      urls: {
        small: '',
        regular: '',
      },
      user: {
        id: '3',
        name: 'test',
        username: 'test',
      },
    },
  ],
  selectedPhoto: null,
  selectedPhotoIndex: null,
  setItems: jest.fn(),
  appendItems: jest.fn(),
  selectPhoto: jest.fn(),
  slideNext: jest.fn(),
  clearSelection: jest.fn(),
};

const search = {
  api: {},
  collections: {},
  searchTerm: '',
  nextPage: 1,
  isLoading: false,
  error: null,
  loadItems: jest.fn(),
  fetchPhotos: jest.fn(),
  searchPhotos: jest.fn(),
  setSearchTerm: jest.fn(),
  setPage: jest.fn(),
  setLoading: jest.fn(),
  setError: jest.fn(),
  recalculateLimit: jest.fn(),
};

describe('Collections Component', () => {
  it('should render and match its own snapshot', () => {
    const component = (
      <Provider search={search} collections={collections}>
        <Collections />
      </Provider>
    );
    const wrapper = mount(component);
    const tree = renderer.create(component).toJSON();

    expect(wrapper).toHaveLength(1);
    expect(tree).toMatchSnapshot();

    expect(wrapper.props()).toHaveProperty('search', search);
    expect(wrapper.props()).toHaveProperty('collections', collections);
  });
  it('should render Views and Preview components', () => {
    const component = (
      <Provider search={search} collections={collections}>
        <Collections />
      </Provider>
    );
    const wrapper = mount(component);

    expect(wrapper.find(Views)).toHaveLength(1);
    expect(wrapper.find(Preview)).toHaveLength(1);
  });
});
