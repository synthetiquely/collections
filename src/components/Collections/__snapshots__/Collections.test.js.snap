// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collections Component should render and match its own snapshot 1`] = `
<div>
  <section
    className="css-wcs7yy"
  >
    <div
      className="css-h0j9oa"
      onClick={[Function]}
      tabIndex={0}
    >
      <img
        alt=""
        className="css-eob8a0"
        onError={[Function]}
        onLoad={[Function]}
        src={undefined}
      />
    </div>
  </section>
</div>
`;
